### PR TITLE
New Adapter: Admatic

### PR DESCRIPTION
**Pull request by bakicam**
We added s2s support for Admatic.

Docs PR: https://github.com/prebid/prebid.github.io/pull/5232

### CODE REVIEW

**Comment by muuki88** [comment]
Docs: https://github.com/prebid/prebid.github.io/pull/5232

File: static/bidder-info/admatic.yaml
Comment (SyntaxNode) [inline static/bidder-info/admatic.yaml:1]: Adapters are not allowed to use an endpoint passed through from the request.

Please see the [Module Rules](https://docs.prebid.org/dev-docs/module-rules.html).  This is listed as Rule 7 in the "Strictly Enforced Rules" section under "Global Module Rules".

File: static/bidder-info/admatic.yaml
Comment (SyntaxNode) [inline static/bidder-info/admatic.yaml:19]: Nitpick: Copy/paste artifact left behind from adtarget.

**Review by SyntaxNode — CHANGES_REQUESTED** [review]

**Comment by onkarvhanumante** [comment]
@bakicam 

consider syning  [Admatic-Sofware:master](https://github.com/Admatic-Sofware/prebid-server/tree/master) branch with [prebid:master](https://github.com/prebid/prebid-server/tree/master) to resolve  PR merge conflicts

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [inline adapters/admatic/admatic.go:37]: nitpick follow sentence case for `Host`

`Host` -> `host`

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [inline adapters/admatic/admatic.go:57]: `host` is defined as required param in `static/bidder-params/admatic.json` 

Prebid server core framework will ensure that admaticExt.Host is not empty. Therefore, `if admaticExt.Host != "" {` check can be removed

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [inline adapters/admatic/admatic.go:49]: nitpick: typo `Hos` should be `host`

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [inline adapters/admatic/admatic.go:54]: nitpick: typo Hos should be host

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [reply adapters/admatic/admatic.go:57]: @bakicam based on adapter code, expectation here is  that all imps should have same host value. 

Should explictily mentioned this behaviour in bidder docs so that publishers are aware of this requirement,

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [inline adapters/admatic/admatic.go:122]: could make use of response util here

https://github.com/prebid/prebid-server/blob/ec729e6e5d50a1306d76332fda52108b9d313c1d/adapters/response.go#L10-L28

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [inline adapters/admatic/admatic.go:129]: by default line 130 will initialize `bidResponse.Currency` as `USD`

reassigning `bidResponse.Currency` on line 131  without checking  `response.Cur` length may overwrite `bidResponse.Currency` with blank value

code block can be refactored as below

```
	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
        if len(response.Cur) != 0 {
	    bidResponse.Currency = response.Cur
        }
```

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [reply adapters/admatic/admatic.go:145]: @bakicam  consider using `MType` to determine media type for bid

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [inline adapters/admatic/admatic.go:154]: should return error if none of the above cases match for determining bid type

File: adapters/admatic/admatic_test.go
Comment (onkarvhanumante) [inline adapters/admatic/admatic_test.go:13]: should use same endpoint mentioned in bidder-info yaml i.e `"http://pbs.admatic.com.tr/?host={{.Host}}"`

this ensures url resolve functions works correctly for endpoint mentioned in bidder-info yaml

**Comment by onkarvhanumante** [comment]
@bakicam  should update json test 
- to have request with multiple imps
- to have multi-format imp request

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [inline adapters/admatic/admatic.go:49]: code path is not tested. should add supplemental tests for covering above code path

File: adapters/admatic/admatic.go
Comment (onkarvhanumante) [inline adapters/admatic/admatic.go:120]: code path is not tested. consider adding supplemental json test for code coverage

File: static/bidder-info/admatic.yaml
Comment (onkarvhanumante) [inline static/bidder-info/admatic.yaml:5]: @bakicam gvl id seems incorrect

```
curl https://vendor-list.consensu.org/v2/vendor-list.json | jq '.vendors."1281"'

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  430k  100  430k    0     0   607k      0 --:--:-- --:--:-- --:--:--  606k
null
```

File: static/bidder-info/admatic.yaml
Comment (onkarvhanumante) [inline static/bidder-info/admatic.yaml:4]: @bakicam  prebid expects maintainer email to be an organization email or mail group but not an email of individual contributor.

`abdulbaki.cam@admatic.com.tr` seems to an email of individual contributor of admatic organization

File: static/bidder-info/admatic.yaml
Comment (onkarvhanumante) [inline static/bidder-info/admatic.yaml:1]: @bakicam  received 404 response for above endpoint. @bakicam  could you confirm if POST endpoint exist for above endpoint url

```
oh0387@vpn-10-249-108-112 ~ % curl -i --location --request POST 'http://pbs.admatic.com.tr/?host='
HTTP/1.1 404 Not Found
Date: Fri, 05 Apr 2024 11:11:00 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 70
Connection: keep-alive
X-Powered-By: Express
ETag: W/"46-wd7cp2ZuKnXKZ3zACcKiNOD1w60"
working-on: admatic-k8s
CF-Cache-Status: DYNAMIC
...

{"message":"Cannot POST /?host=","error":"Not Found","statusCode":404}%
```

**Comment by onkarvhanumante** [comment]
@bakicam  should create and link PR for bidder doc in https://github.com/prebid/prebid.github.io repository

File: adapters/admatic/admatic.go
Comment (bretg) [reply adapters/admatic/admatic.go:54]: You must have a default Host value. It is not reasonable to expect global publishers to figure out which host to call. Please see https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#consider-your-geography

File: static/bidder-info/admatic.yaml
Comment (bakicam) [reply static/bidder-info/admatic.yaml:5]: Hello, I can see it's defined here. https://vendor-list.consensu.org/v3/vendor-list.json

File: adapters/admatic/admatic.go
Comment (bakicam) [reply adapters/admatic/admatic.go:54]: Hello, the user does not need to define the host name here, the host value here is the publisher id that we will give to the user.

**Comment by bakicam** [comment]
https://github.com/prebid/prebid-server/pull/3608#issuecomment-2039539996

Hello,
PR is here https://github.com/prebid/prebid.github.io/pull/5232

**Comment by onkarvhanumante** [comment]
@bakicam PR has some merge conflicts. Requesting to resolve them.

File: static/bidder-info/admatic.yaml
Comment (onkarvhanumante) [reply static/bidder-info/admatic.yaml]: @bakicam  any feedback on above comment

File: static/bidder-info/admatic.yaml
Comment (onkarvhanumante) [reply static/bidder-info/admatic.yaml]: @bakicam  any feedback on above comment

File: static/bidder-info/admatic.yaml
Comment (farukcam) [reply static/bidder-info/admatic.yaml]: Hello, we fixed this problem

File: static/bidder-info/admatic.yaml
Comment (farukcam) [reply static/bidder-info/admatic.yaml]: Merhaba,
Buradaki committe değişiklikler yapıldı. https://github.com/prebid/prebid-server/pull/3608/commits/f98a6490e6053b5206b233be1d50a3310706880b

**Comment by farukcam** [comment]
https://github.com/prebid/prebid-server/pull/3608#issuecomment-2058751261 Hello, we have made arrangements for this.

**Comment by farukcam** [comment]
Hello,
Is there any progress regarding our request?

**Comment by onkarvhanumante** [comment]
> Hello, Is there any progress regarding our request?

@bakicam  requesting to push updates to resolve merge conflicts

**Comment by bakicam** [comment]
> > Hello, Is there any progress regarding our request?
> 
> @bakicam requesting to push updates to resolve merge conflicts

Hello,
We are currently just waiting for approval of SyntaxNode solutions. I would like you to check it.

**Comment by farukcam** [comment]
Hello,
We want your support

**Comment by onkarvhanumante** [comment]
@bakicam  @farukcam  PR has some merge conflicts. PTAL to resolve them. Requesting to sync [Admatic-Sofware:master](https://github.com/Admatic-Sofware/prebid-server/tree/master) with [prebid:master](https://github.com/prebid/prebid-server/tree/master) changes

**Comment by farukcam** [comment]
@onkarvhanumante I synced

**Comment by farukcam** [comment]
@onkarvhanumante @Sonali-More-Xandr Hello, I think there is a problem here. It has been merged but we cannot see the adapter.

**Comment by onkarvhanumante** [comment]
image

image



@farukcam changes [force-pushed](https://github.com/prebid/prebid-server/compare/109984445036fb33a769d909cda7ab8d18bfcb1c..bc89164de1ea415ccc8d6ff2b5aacdea161e04f2)  resulted into discarding file changes.

**Comment by farukcam** [comment]
@onkarvhanumante I'm so sorry, I think we made a mistake. Should we create a new job request?

**Comment by onkarvhanumante** [comment]
> @onkarvhanumante I'm so sorry, I think we made a mistake. Should we create a new job request?

yes please create new PR

**Comment by farukcam** [comment]
@onkarvhanumante I request you to close this https://github.com/prebid/prebid-server-java/issues/3165

**Comment by onkarvhanumante** [comment]
> @onkarvhanumante I request you to close this [prebid/prebid-server-java#3165](https://github.com/prebid/prebid-server-java/issues/3165)

I don't have privledge to close PR from prebid-server-java repo. Added this https://github.com/prebid/prebid-server-java/issues/3165#issuecomment-2089907868 for closing PR.

**Comment by farukcam** [comment]
@onkarvhanumante Hello, We created a new application here https://github.com/prebid/prebid-server/pull/3654. Thank you for your support.

**Pull request merged**
